### PR TITLE
fixing root cause of SI-6863

### PR DESCRIPTION
**Pull request by magarciaEPFL**
review by @JamesIry

### CODE REVIEW

**Comment by retronym** [comment]
Could you please update the commit comment in the style of https://github.com/scala/scala/blob/master/CONTRIBUTING.md / http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html ?

```
SI-1234 Short commit summary.

Lines here word wrapped to 72 chars.

 - Bullet points like this, if used.
```

I realise that this doesn't directly fix that referenced ticket, but I would still put the ticket number at the start of the message.

File: src/compiler/scala/tools/nsc/transform/LambdaLift.scala [hunk 1]
```diff
@@ -32,6 +32,21 @@ abstract class LambdaLift extends InfoTransform {
...
   }
 
+  /** scala.runtime.*Ref classes */
+  private lazy val allRefClasses: Set[Symbol] = {
+    refClass.values.toSet ++ volatileRefClass.values.toSet ++ Set(VolatileObjectRefClass, ObjectRefClass)
+  }
+
+  /** Each scala.runtime.*Ref class has a static method `create(value)` that simply instantiates the Ref to carry that value. */
+  private lazy val refCreateMethod: Map[Symbol, Symbol] = {
+    allRefClasses map (x => x -> getMemberMethod(x.companionModule, nme.create)) toMap
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala:42]: A touch more compact:

``` scala
mapFrom(allRefClasses.toList)(x => getMemberMethod(x.companionModule, nme.create))
```

File: src/compiler/scala/tools/nsc/transform/LambdaLift.scala [hunk 2]
```diff
@@ -444,56 +459,21 @@ abstract class LambdaLift extends InfoTransform {
...
-                val tempSym = currentOwner.newValue(unit.freshTermName("temp"), expr.pos) setInfo expr.tpe
-                val tempDef = ValDef(tempSym, expr) setPos expr.pos
-                val tempRef = Ident(tempSym) setPos expr.pos
-                Block(tempDef, New(sym.tpe, tempRef))
-            })
-            def refConstrCase(cdef: CaseDef): CaseDef =
-              CaseDef(cdef.pat, cdef.guard, refConstr(cdef.body))
-
-            treeCopy.ValDef(tree, mods, name, tpt1, refConstr(constructorArg))
+            treeCopy.ValDef(tree, mods, name, tpt1, factoryCall)
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala:476]: Lovely!

**Comment by magarciaEPFL** [comment]
Reworked PR submitted.

**Pull request closed without merging**